KAFKA-3214: Added system tests for compressed topics

apovzner opened the pull request:
Added CompressionTest that tests 4 producers, each using a different compression type and one not using compression.

Enabled VerifiableProducer to run producers with different compression types (passed in the constructor). This includes enabling each producer to output unique values, so that the verification process in ProduceConsumeValidateTest is correct (counts acks from all producers). 

Also a fix for console consumer to raise an exception if it sees the incorrect consumer output (before we swallowed an exception, so was hard to debug the issue).

apovzner committed:
KAFKA-3214: Added system tests for compressed topics

hachikuji commented on tests/kafkatest/sanity_checks/test_verifiable_producer.py:
```diff
@@ -36,21 +36,25 @@ def __init__(self, test_context):
-    @parametrize(producer_version=str(TRUNK))
-    def test_simple_run(self, producer_version=TRUNK):
+
+    @parametrize(producer_version=str(LATEST_0_8_2), compression_type=None)
```
Not related to this patch, but I wonder if we should also be including a LATEST_0_9_0?

apovzner replied:
Yes, this will be added when I implement upgrade test (0.9 to 0.10).

hachikuji commented on tests/kafkatest/sanity_checks/test_verifiable_producer.py:
```diff
@@ -36,21 +36,25 @@ def __init__(self, test_context):
+
+    @parametrize(producer_version=str(LATEST_0_8_2), compression_type=None)
+    @parametrize(producer_version=str(TRUNK), compression_type=None)
+    @parametrize(producer_version=str(TRUNK), compression_type=["snappy"])
```
Any particular reason to test compression on this test if we have the test cases below?

apovzner replied:
The reasoning is that this is a smoke test, and compression I guess is one of the basic behaviors. I don't have strong opinion about this. @granders Do you think it makes sense to do smoke test for compression if we have a separate produce -> consume -> verify system tests for compressed topics?

apovzner replied:
I thought about this more, and the new compression test is just produce -> consume -> validate, so does not seem to make sense to also add a smoke test. I will remove this test.

hachikuji commented on tests/kafkatest/tests/compression_test.py:
```diff
@@ -0,0 +1,79 @@
+
+        self.topic = "test_topic"
+        self.zk = ZookeeperService(test_context, num_nodes=1)
+        self.kafka = KafkaService(test_context, num_nodes=3, zk=self.zk, topics={self.topic: {
```
Are multiple brokers necessary for a compression test?

apovzner replied:
You are right, probably not. I will do one broker.

hachikuji commented on tests/kafkatest/tests/compression_test.py:
```diff
@@ -0,0 +1,79 @@
+        return super(CompressionTest, self).min_cluster_size() + self.num_producers + self.num_consumers
+
+    @parametrize(compression_types=["snappy","gzip","lz4"])
+    @parametrize(compression_types=["snappy","none"])
```
This parameterization is a little confusing. Could we not just do this?

``` python
@parametrize(compression_types=["none", "snappy", "gzip", "lz4"])
```

apovzner replied:
The idea was to do one test with each of three producers using different compression, and then first and third producers using snappy and second producer is not using compression. As we just discussed, I will do 4 producers, and the 4th will not use compression; does not seem to add value to do that in two separate tests.

hachikuji replied:
Makes sense, thanks.

hachikuji commented on tests/kafkatest/services/verifiable_producer.py:
```diff
@@ -56,20 +66,25 @@ def __init__(self, context, num_nodes, kafka, topic, max_messages=-1, throughput
         node.account.create_file(VerifiableProducer.LOG4J_CONFIG, log_config)
 
         # Create and upload config file
+        if self.compression_types is not None:
```
Not sure I understand why this is a list. Do we ever test with more than one compression type?

hachikuji replied:
An alternative might be to let verifiable producer accept just one compression type. Then in your test case, you could create separate instances, each with a different compression type. Seems a little more intuitive that way to me.

apovzner replied:
This will require pretty big change/rewrite to ProduceConsumeValidateTest. It currently assumes one VerifiableProducer. Let me just write a comment above this line.

hachikuji replied:
Yeah, good point. It's just a little weird that this is the only configuration setting which is allowed to vary (the fact that we need a big comment to explain it kind of suggests as much). But I suppose that there won't be many use cases for compression, so maybe it's ok.

hachikuji commented on tests/kafkatest/tests/compression_test.py:
```diff
@@ -0,0 +1,79 @@
+
+class CompressionTest(ProduceConsumeValidateTest):
+    """
+    These tests validate partition reassignment.
```
Looks like this comment needs updating.

apovzner committed:
KAFKA-3214: removed compression smoke test; merged compression test into one

apovzner committed:
revertign back to no changes in test verifiable producer

apovzner committed:
Using one broker for compression tests

apovzner commented:
The new system test passes locally and on Jenkins.

hachikuji commented:
LGTM

ijuma commented:
Nice @apovzner, thanks for improving our system tests coverage!

apovzner committed:
Added optional prefix to verifiable producer to ensure that producers produce unique values in multiproducer test

apovzner commented:
re-run impacted system tests, they passed.

granders commented on tests/kafkatest/tests/compression_test.py:
```diff
@@ -0,0 +1,78 @@
+
+        self.kafka.security_protocol = "PLAINTEXT"
+        self.kafka.interbroker_security_protocol = self.kafka.security_protocol
+        new_consumer = False if  self.kafka.security_protocol == "PLAINTEXT" else True
```
Is there any reason to ever use old consumer here?

hachikuji replied:
Until the old consumer is deprecated, would it make sense to run this test against both consumers? I think the client implementations are different enough that there would be value in ensuring that compression works with both.

apovzner replied:
@hachikuji so basically have new_consumer as parameter, since we are going to always run this test with "PLAINTEXT" security protocol?

hachikuji replied:
Yeah, that would make sense to me.

granders commented on tests/kafkatest/tests/compression_test.py:
```diff
@@ -0,0 +1,78 @@
+                                        new_consumer=new_consumer, consumer_timeout_ms=60000, message_validator=is_int_with_prefix)
+        self.kafka.start()
+        
+        self.run_produce_consume_validate()
```
If you run this with no args, all this will do is start the producers/consumer, and then immediately stop. Some small number of messages are probably produced, but you have no guarantee over how many have been acked before the clients are stopped.

I think it may make sense to ensure that a known number of messages have been acknowledged.

So you could:
a) In VerifiableProducer, specify for example 1000 messages per producer
b) something like `run_produce_consume_validate(lambda _: wait_until(self.producer.num_acked >= 4000, timeout_sec=30, err_message))`

This will mean each producer produces a known number of messages.
Also, we expect each to be acked since there is no failure injection, so this enforces the fact that we expect all messages to be acked

apovzner replied:
Oh good to know. I will do that. When I ran the test, it did produce about 5000 message. But agree it is better to make it more predictable.

granders commented on tests/kafkatest/services/verifiable_producer.py:
```diff
@@ -87,10 +109,10 @@ def _worker(self, idx, node):
                     if data["name"] == "producer_send_error":
                         data["node"] = idx
-                        self.not_acked_values.append(int(data["value"]))
+                        self.not_acked_values.append(self.message_validator(data["value"]))
```
Now that `message_validator` methods are being used here in VerifiableProducer, it is odd that with the existing validators the behavior is now:
If valid, append original value (possibly parsed into an int)
else, append None

That actually seems pretty surprising. Previously, if the value was not parseable, an error would be thrown immediately.

apovzner replied:
@granders Very good point. Before, if the format was wrong, int() would throw an exception. As we just discussed, the consumer side currently swallows the exception -- but if we got a message with unexpected format we should fail the test. I will change both validators (on both producer and consumer side) to throw an exception if format is wrong.

apovzner replied:
Done. Made both validators too raise exception if wrong format.

apovzner committed:
Enforce minimum produced messages for compression test

apovzner commented:
@hachikuji and @granders I addressed your comments. Now compression test tests both new and old consumer. Compression test runs until each producer produces at least 1000 messages.
Let me know if you have any more comments.

granders commented on tests/kafkatest/services/verifiable_producer.py:
```diff
@@ -166,6 +195,12 @@ def num_not_acked(self):
             return len(self.not_acked_values)
 
+    def each_produced_at_least(self, count):
+        for idx in range(1, self.num_nodes):
```
Should this be protected by `with self.lock:`?

apovzner committed:
added lock around checking number of produced messages in verifiable producer

granders commented:
@apovzner Thanks for the updates. LGTM

gwenshap commented:
LGTM.
